In [1]:
import math

def get_range(data):
    return max(data) - min(data)

def get_mean(data):
    return sum(data) / len(data)

def get_variance(data, sample=False):
    m = get_mean(data)
    squared_diffs = [(x - m) ** 2 for x in data]
    divisor = len(data) - 1 if sample else len(data)
    return sum(squared_diffs) / divisor

def get_std_dev(data, sample=False):
    return math.sqrt(get_variance(data, sample))

def get_cv(data):
    return (get_std_dev(data) / get_mean(data)) * 100

## Task 1 — Range of Daily Step Counts

In [1]:
steps = [6500, 7200, 8000, 6000, 9000, 7500, 8200]

step_range = get_range(steps)
print('Daily step counts:', steps)
print('Range =', max(steps), '-', min(steps), '=', step_range, 'steps')

# **Interpretation:** The range is **3000 steps**, spanning a low of 6000 (day 4) to a high of 9000 (day 5).
# That's a wide swing for one week — the user isn't walking a consistent amount each day, 
# likely mixing sedentary days with much more active ones. Range only looks at the two extreme values though, 
# so it can be skewed by a single unusually high or low day; it doesn't say anything about how the other 5 days behaved.

Daily step counts: [6500, 7200, 8000, 6000, 9000, 7500, 8200]
Range = 9000 - 6000 = 3000 steps


## Task 2 — Variance & Standard Deviation of Zomato Order Amounts

In [1]:
orders = [250, 400, 350, 300, 500]

order_mean = get_mean(orders)
order_variance = get_variance(orders)
order_std = get_std_dev(orders)

print('Order amounts:', orders)
print('Mean = Rs.', round(order_mean, 2))
print('Variance (population) =', round(order_variance, 2))
print('Standard Deviation (population) =', round(order_std, 2))


# Note on formula choice: 
# this uses the population formula (divide by n), treating all 5 orders as the complete dataset rather than a sample. 
# If you instead used pandas' .var() / .std(), they default to the sample formula (divide by n-1) and would return 9250.0 and 96.18 instead — 
# worth checking which one an assignment or tool expects before comparing numbers across sources.

Order amounts: [250, 400, 350, 300, 500]
Mean = Rs. 360.0
Variance (population) = 7400.0
Standard Deviation (population) = 86.02


## Task 3 — Coefficient of Variation of Mobile Data Usage

In [1]:
data_usage = [3.2, 2.8, 3.5, 3.0, 2.9, 3.1]

usage_mean = get_mean(data_usage)
usage_std = get_std_dev(data_usage)
usage_cv = get_cv(data_usage)

print('Monthly data usage (GB):', data_usage)
print('Mean =', round(usage_mean, 2), 'GB')
print('Standard Deviation =', round(usage_std, 2), 'GB')
print('Coefficient of Variation =', round(usage_cv, 2), '%')

Monthly data usage (GB): [3.2, 2.8, 3.5, 3.0, 2.9, 3.1]
Mean = 3.08 GB
Standard Deviation = 0.23 GB
Coefficient of Variation = 7.35 %


**Interpretation:** CV ≈ **7.35%**. As a rule of thumb, a CV under ~15% signals low, consistent variability. This user's data usage is stable and predictable — it stays close to 3 GB every month with only minor swings, not the pattern of someone whose usage spikes unpredictably.

## Task 4 — Standard Deviation of Instagram Post Likes

In [1]:
likes = [120, 150, 130, 110, 170, 140, 135, 160, 125, 155]

likes_mean = get_mean(likes)
likes_std = get_std_dev(likes)
likes_cv = get_cv(likes)

print('Instagram post likes:', likes)
print('Mean =', round(likes_mean, 2))
print('Standard Deviation =', round(likes_std, 2))
print('CV =', round(likes_cv, 2), '%')

Instagram post likes: [120, 150, 130, 110, 170, 140, 135, 160, 125, 155]
Mean = 139.5
Standard Deviation = 18.09
CV = 12.97 %


**Interpretation:** The standard deviation (~18.09 likes) is about **13% of the mean** (139.5 likes) — a moderate spread, not tight and not chaotic. A couple of posts (170 and 110 likes) pull noticeably away from average, so engagement is reasonably consistent overall but still sensitive to specific posts — content type, timing, or format is probably driving those outliers.

## Task 5 — Variance & SD of an AI-Generated UPI Transaction Dataset

Dataset below represents 8 plausible daily UPI transaction amounts (₹) for a week, generated to satisfy the task's 'at least 8 values' requirement.

In [1]:
upi_transactions = [450, 1200, 300, 800, 950, 400, 1500, 650]

upi_mean = get_mean(upi_transactions)
upi_variance = get_variance(upi_transactions)
upi_std = get_std_dev(upi_transactions)
upi_cv = get_cv(upi_transactions)

print('UPI transaction amounts (Rs.):', upi_transactions)
print('Mean = Rs.', round(upi_mean, 2))
print('Variance =', round(upi_variance, 2))
print('Standard Deviation = Rs.', round(upi_std, 2))
print('CV =', round(upi_cv, 2), '%')

UPI transaction amounts (Rs.): [450, 1200, 300, 800, 950, 400, 1500, 650]
Mean = Rs. 781.25
Variance = 153085.94
Standard Deviation = Rs. 391.26
CV = 50.08 %


**Interpretation:** CV ≈ **50.08%** — much higher than the mobile data or likes examples. Spending swings widely day to day, from ₹300 up to ₹1500. This isn't stable, routine spending; it looks more like a mix of small everyday purchases plus one or two larger one-off payments, and those large values are what pull the variance up so sharply.